fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/lowcardinality/DecodeCollector.java

```diff
@@ -0,0 +1,609 @@
+                disableColumns.union(agg.getUsedColumns());
+                disableColumns.union(key);
+                continue;
+            }
+            if (agg.getChildren().size() != 1 || !agg.getChildren().get(0).isColumnRef()) {
+                disableColumns.union(agg.getUsedColumns());
+                disableColumns.union(key);
+            }
+        }
+
+        if (!disableColumns.isEmpty()) {
+            info.decodeStringColumns.union(info.inputStringColumns);
+            info.decodeStringColumns.intersect(disableColumns);
+            info.inputStringColumns.except(info.decodeStringColumns);
+        }
+
+        info.outputStringColumns.clear();
+        for (ColumnRefOperator key : aggregate.getAggregations().keySet()) {
+            if (disableColumns.contains(key)) {
+                continue;
+            }
+            CallOperator value = aggregate.getAggregations().get(key);
+            if (!info.inputStringColumns.containsAll(value.getUsedColumns())) {
+                continue;
+            }
+            // aggregate ref -> aggregate expr
+            stringAggregateExpressions.computeIfAbsent(key.getId(), x -> Lists.newArrayList()).add(value);
+            // min/max should replace to dict column, count/count distinct don't need
+            if (FunctionSet.MAX.equals(value.getFnName()) || FunctionSet.MIN.equals(value.getFnName())) {
+                info.outputStringColumns.union(key.getId());
+                stringRefToDefineExprMap.putIfAbsent(key.getId(), value);
+                expressionStringRefCounter.put(key.getId(), 1);
+            } else if (aggregate.getType().isLocal() || aggregate.getType().isDistinctLocal()) {
+                // count/count distinct, need output dict-set in 1st stage
+                info.outputStringColumns.union(key.getId());
+            }
+        }
+
+        for (ColumnRefOperator groupBy : aggregate.getGroupBys()) {
+            if (info.inputStringColumns.contains(groupBy) && !info.decodeStringColumns.contains(groupBy)) {
+                info.outputStringColumns.union(groupBy);
+            }
+        }
+
+        for (ColumnRefOperator partition : aggregate.getPartitionByColumns()) {
+            if (info.inputStringColumns.contains(partition) && !info.decodeStringColumns.contains(partition)) {
+                info.outputStringColumns.union(partition);
+            }
+        }
+
+        return info;
+    }
+
+    @Override
+    public DecodeNodeInfo visitPhysicalDistribution(OptExpression optExpression, DecodeNodeInfo context) {
+        if (context.outputStringColumns.isEmpty()) {
+            return DecodeNodeInfo.EMPTY;
+        }
+        DecodeNodeInfo result = context.createOutputInfo();
+        // join on-predicate don't support low-cardinality optimization, must decode before shuffle
+        if (context.parent != null && context.parent.getOp() instanceof PhysicalJoinOperator) {
+            return visitPhysicalJoin(context.parent, context);
```

Comment:
What the different when the parent is not a joinOp?

Reply:
JoinOperator on-predicate doesn't support decode-column, and must decode before shuffle, but other node supported